### ISSUE TITLE
Access the underlying buffer object associated with a texture.

### ISSUE THREAD

**Issue by nitronoid** [google/filament#1299]
**Is your feature request related to a problem? Please describe.**
I would like to access the buffer handle associated with a `filament::Texture`, so that I can use the cuda interop api and write directly to the buffer from the GPU.

**Describe the solution you'd like**
From the little investigation I've done, I can see that the derived `FTexture` class holds the `HwTexture` handle, and  the `backend::Handle` api seems to offer access to the underlying buffer handle. I'm wondering if there's a "clean" way for me to access that handle, without casting and including private headers? If this isn't supported/isn't planned, what is the best/safest work around I could use?

Thanks

**Comment by romainguy** [comment]
By buffer object do you mean you want access to the OpenGL texture id? One of the reasons we don't expose things like this is to be able to transparently support multiple backends. Can you do it the other way around and use a shared GL context to give Filament the texture you want to share with CUDA? That we support (even though it's unfortunately GL specific).

**Comment by nitronoid** [comment]
Yes the id is what I'm after. Hm I wasn't aware of the other option, is there something in the samples similar to what you suggest?

**Comment by romainguy** [comment]
Unfortunately no but what you're after is:

https://github.com/google/filament/blob/master/filament/include/filament/Engine.h#L185

It only works with EGL though, I don't know if you can use this on desktop :/

@pixelflinger any thoughts on this feature request?

**Comment by prideout** [comment]
(Note that a very similar feature request has been made from a team within Google.)

**Comment by nitronoid** [comment]
@romainguy That is unfortunate :( 

As for the backend issue, I'm not very familiar with vulkan but from the examples I've seen, writing to a `VkBuffer` would require these steps: 
1. Get the memory handle to the `VkBuffer` with a call to whatever gets returned by this `vkGetInstanceProcAddr(instance, "vkGetMemoryFdKHR")`
2. Import and map the memory with cuda `cudaImportExternalMemory` and `cudaExternalMemoryGetMappedBuffer`
3. Block vulkan using a sephamore
4. Execute CUDA kernel that writes to the `VkBuffer` using the memory handle from step 1
5. Unblock vulkan
6. unmap the buffer

So the only thing I really need from vulkan is that memory handle (I think) so that I can map it. Can I ask, what does the `backend::Handle<HwTexture>` store when the backend is Vulkan?

And with regards to metal, I'm clueless there...

**Comment by romainguy** [comment]
`HwTexture` is a `VulkanTexture` which contains a `VkImage`, a `VkImageView` and a `VkDeviceMemory`. @prideout can say more about this since he wrote the Vulkan backend.

**Comment by prideout** [comment]
You would also need a device, command buffer, etc...  Long term I think we should support sharing in a more first-class way using platform-specific mechanisms (e.g. AHardwareBuffer), and in the short term expose a small OpenGL-specific mechanism.

@nitronoid would you be unblocked if we provided an OpenGL-only solution?

**Comment by nitronoid** [comment]
@prideout Yes an OGL solution would be perfect! 
I was just speaking hypothetically about the Vulkan part out of interest, however having access to the OpenGL texture ID would be much more useful right now. Thanks!

**Comment by romainguy** [comment]
I imagine the texture ID isn't enough and you also need access to the GL context. It's the kind of stuff I'd really like to avoid exposing.

**Comment by nitronoid** [comment]
@romainguy  Yes I would need that too. Is there any way to get the shared context working on desktop?

**Comment by nitronoid** [comment]
Actually, I only really need the context so that I can ensure it is the current context for my thread. Would it be possible to offer some sort of `makeCurrent` function so I can do that in a platform agnostic way? 
Then I could just call `makeCurrent`, and get the now valid texure Id somehow.

I see that the private implementation of `SwapChain` has something similar, but requires the backend as a parameter.

**Comment by romainguy** [comment]
So you are touching one the #1 reason why we use a separate thread for OpenGL: we do not want the app to ever be able to do anything that might modify the state of Filament's GL context (we do careful state tracking, etc.). That's also why we never expose APIs that let run user code on the driver thread.

**Comment by nitronoid** [comment]
I see, not sure where to go from here then. Unless you implement the mapping of the texture with the cudaInterop calls internally (which is a bit of a stretch), I'm not sure how I could do it without running on the driver thread :/

How would you prevent the user from running code on the driver thread when using that sharedContext?

**Comment by romainguy** [comment]
The shared context doesn't let you run arbitrary thread in our driver (by driver I mean Filament's backend, not the GPU driver of your system). It does allow you to do a few naughty things (you could delete some of Filament's resources for instance) but it's somewhat limited. The best solution would be to look into the equivalent of EGL's shared context for WGL/XGL/etc. What platform are you targeting?

**Comment by nitronoid** [comment]
RHEL-7 desktop with Nvidia Quadro cards. I think that's GLX?

**Comment by romainguy** [comment]
Yes, we use GLX there. It's apparently trivial to [enable context sharing](https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glXCreateContext.xml). We just need to pass the `void* sharedContext` you give us to `glXCreateContext`.

@prideout Any chance you could give this a try?

**Comment by nitronoid** [comment]
https://github.com/google/filament/blob/5b4ec6fd7d5b1ed1bef14f63c06eccddb6a55b72/filament/backend/src/opengl/PlatformGLX.cpp#L136-L137

Seems like it's already being passed? Unless I'm mistaken.

**Comment by romainguy** [comment]
Look at that, apparently we already took care of it :) Give it a try and let us know how it goes then!

**Comment by nitronoid** [comment]
Great thanks for helping me out!
Final question: I'm guessing there's no way I can convert an OpenGL texture I create on the shared context with `glGenTextures` into a `filament::Texture`? That way I could keep using your nice API, and have the texture Id.

**Comment by romainguy** [comment]
The way to make it work is to create a `Stream` with an external texture id (https://github.com/google/filament/blob/master/filament/include/filament/Stream.h). The stream can then be attached to a `Texture`. I believe you can ignore the comment about `GL_TEXTURE_EXTERNAL_OES` (@pixelflinger is the doc wrong here? Texture external should be for `SurfaceTexture` not shared textures right?).

**Comment by nitronoid** [comment]
Sorry for no updates, I've been handling another task. I will get back to this and test it soon if you don't mind leaving the issue open a little longer?

**Comment by romainguy** [comment]
Take your time :)
